### PR TITLE
Only allow users with the CAS3_REFERRER role to interact with Apply / Only allow users with the CAS3_ASSESSOR role to interact with Manage

### PR DESCRIPTION
**Pull request by bryangaledxw**
# Changes in this PR

* This PR restricts manager routes to user with the assessor role, and apply routes to users with the referrer role
* When the user first reaches the service, they will be redirected to either the manage dashboard, or the apply start page

# Release checklist

[Release process documentation](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/edit-v2/4247847062?draftShareId=a1c360ab-bd31-4db1-aae3-7cc002761de9).

## Pre-merge checklist

- [ ] Are any changes required to dependent services for this change to work?
  (eg. CAS API)
- [ ] Have they been released to production already?

## Post-merge checklist

- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to test
- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to preprod
- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to prod

### CODE REVIEW

**Comment by bryangaledxw** [comment]
[Only allow users with the CAS3_ASSESSOR role to interact with Manage](https://trello.com/c/209Srqfj/1337-only-allow-users-with-the-cas3assessor-role-to-interact-with-manage)

**Comment by bryangaledxw** [comment]
[Only allow users with the CAS3_REFERRER role to interact with Apply](https://trello.com/c/jJWy1iBY/1285-only-allow-users-with-the-cas3referrer-role-to-interact-with-apply)

**Comment by bryangaledxw** [comment]
[Only allow users with the CAS3_REFERRER role to interact with Apply](https://trello.com/c/jJWy1iBY/1285-only-allow-users-with-the-cas3referrer-role-to-interact-with-apply)

File: server/middleware/roleMiddleware.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+import { UnauthorizedError } from '../utils/errors'
+
+export const createRoleMiddleware = (role: UserRole) => {
+  return (handler: RequestHandler) => wrapHandler(handler, role)
+}
+
+const wrapHandler =
+  (handler: RequestHandler, role: UserRole) => async (req: Request, res: Response, next: NextFunction) => {
+    if (res.locals.user.roles.includes(role)) {
+      await handler(req, res, next)
```
Comment (Harriethw) [inline server/middleware/roleMiddleware.ts:12]: minor: the IDE is telling me that `await` has no effect on this function, but haven't got it running to tell locally :)

**Review by Harriethw — APPROVED** [review]
nice one! definitely bookmarking this in case we have to implement something similar :)

File: server/middleware/roleMiddleware.ts (hunk 1 above)
Comment (bryangaledxw) [reply server/middleware/roleMiddleware.ts:12]: Thanks for the heads up on this - took another look at the code and realised the audit step is happening before the role step, which wasn't what I intended - will need to take another look at things

**Review by tahb — APPROVED** [review]
LGTM

**Pull request merged**